instantlaunches/users.go

```diff
@@ -0,0 +1,287 @@
+package instantlaunches
+
+import (
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+
+	"github.com/labstack/echo/v4"
+)
+
+const userMappingQuery = `
+    SELECT u.id,
+           u.version,
+           u.instant_launches as mapping
+      FROM user_instant_launches u
+      JOIN users ON u.user_id = users.id
+     WHERE users.username = ?
+  ORDER BY u.version DESC
+     LIMIT 1;
+`
+
+const updateUserMappingQuery = `
+    UPDATE ONLY user_instant_launches AS def
+            SET def.instant_launches = jsonb_object(?)
+           FROM users
+          WHERE def.version = (
+              SELECT max(version)
+                FROM user_instant_launches
+          )
+            AND def.user_id = users.id
+            AND users.username = ?
+          RETURNING def.instant_launches;
+`
+
+const deleteUserMappingQuery = `
+	DELETE FROM ONLY user_instant_launches AS def
+	USING users
+	WHERE def.user_id = users.id
+	  AND users.username = ?
+	  AND def.version = (
+		  SELECT max(version)
+		    FROM user_instant_launches
+	  )
+`
+
+const createUserMappingQuery = `
+	  INSERT INTO user_instant_launches (instant_launches, user_id)
+	  VALUES ( ?, (SELECT id FROM users WHERE username = ?) )
+	  RETURNING instant_launches;
+`
+
+// UserMapping returns the user's instant launch mappings.
+func (a *App) UserMapping(user string) (*UserInstantLaunchMapping, error) {
+	m := &UserInstantLaunchMapping{}
+	err := a.DB.Get(m, userMappingQuery, user)
+	return m, err
+}
+
+// UserMappingHandler is the echo handler for the http API that returns the user's
+// instant launch mappings.
+func (a *App) UserMappingHandler(c echo.Context) error {
+	user := c.Param("username")
+	m, err := a.UserMapping(user)
+	if err != nil {
+		return err
+	}
+	return c.JSON(http.StatusOK, m)
+}
+
+// UpdateUserMapping updates the the latest version of the user's custom
+// instant launch mappings.
+func (a *App) UpdateUserMapping(user string, update *InstantLaunchMapping) (*InstantLaunchMapping, error) {
+	marshalled, err := json.Marshal(update)
+	if err != nil {
+		return nil, err
+	}
+	updated := &InstantLaunchMapping{}
+	err = a.DB.QueryRowx(updateUserMappingQuery, marshalled, user).Scan(updated)
+	return updated, err
+}
+
+// UpdateUserMappingHandler is the echo handler for the HTTP API that updates the user's
+// instant launch mapping.
+func (a *App) UpdateUserMappingHandler(c echo.Context) error {
+	user := c.Param("username")
+	newdefaults := &InstantLaunchMapping{}
+	readbytes, err := ioutil.ReadAll(c.Request().Body)
+	if err != nil {
+		return err
+	}
+
+	if err = json.Unmarshal(readbytes, newdefaults); err != nil {
+		return err
+	}
```

Comment:
This probably indicates a malformed request body as well.

Reply:
Should be fixed now, thanks!